parglare/parser.py

```diff
@@ -715,26 +727,34 @@ def _lexical_disambiguation(self, context, tokens):
             h_print("Lexical disambiguation.",
                     " Tokens: {}".format([x for x in tokens]), level=1)
 
+        if len(tokens) <= 1:
+            return tokens
+
+        # prefer STOP over EMPTY and EOF
+        if STOP_token in tokens:
+            tokens = [t for t in tokens if t not in (EMPTY_token, EOF_token)]
+        # prefer EMTPY over EOF
+        if EMPTY_token in tokens:
+            tokens = [t for t in tokens if t != EOF_token]
+
         # Longest-match strategy.
         max_len = max((len(x.value) for x in tokens))
         tokens = [x for x in tokens if len(x.value) == max_len]
         if self.debug:
             h_print("Disambiguation by longest-match strategy.",
                     "Tokens: {}".format([x for x in tokens]), level=1)
         if len(tokens) == 1:
-            return tokens[0]
-        else:
-            # Finally try to find preferred token.
-            pref_tokens = [x for x in tokens if x.symbol.prefer]
-            if len(pref_tokens) == 1:
-                if self.debug:
-                    h_print("Preferring token {}.".format(pref_tokens[0]),
-                            level=1)
-                return pref_tokens[0]
-            elif len(pref_tokens) > 1:
-                tokens = pref_tokens
+            return tokens
 
-        raise DisambiguationError(Location(context), tokens)
+        # try to find preferred token.
+        pref_tokens = [x for x in tokens if x.symbol.prefer]
+        if len(pref_tokens) > 0:
```

Comment:
`if pref_tokens:`

Reply:
too bad pep8 promotes using iterables as bools ;( iterables are not bools. but I won't debate here, but just will adjust to the standard

Reply:
Yup. That's idiomatic Python, like it or not :)